src/libopensc/pkcs15-piv.c

```diff
@@ -1143,7 +1143,56 @@ sc_log(card->ctx,  "DEE Adding pin %d label=%s",i, label);
 
 		strncpy(prkey_obj.label, prkeys[i].label, SC_PKCS15_MAX_LABEL_SIZE - 1);
 		prkey_obj.flags = prkeys[i].obj_flags;
-		prkey_obj.user_consent = prkeys[i].user_consent; /* only Sign key */
+
+		/* If Yubikey, try to get pin policy from ATTESTATION certificate */
+		/* TODO, for now, only check to turn off user_consent on Sign key */
+		/* could look at every key, to set piv and touch policy */
+
+		prkey_obj.user_consent = prkeys[i].user_consent; /* set the PIV default */
+		if (prkeys[i].user_consent) {
+			sc_cardctl_piv_yubico_attestation_cert_t attcertctl;
+
+			memset(&attcertctl, 0, sizeof(attcertctl));
+			attcertctl.key = prkeys[i].ref;
+
+			r = sc_card_ctl(card, SC_CARDCTL_PIV_YUBICO_ATTESTATION_CERT, &attcertctl);
+			if (r < 0 || attcertctl.der.value == NULL || attcertctl.der.len == 0) {
+				sc_log(card->ctx," SC_CARDCTL_PIV_YUBICO_ATTESTATION_CERT failed r=%d", r);
+			} else {
+				struct sc_object_id pp_oid = {{1, 3, 6, 1, 4, 1, 41482, 3, 8, -1}}; /* pin policy extension */
+				sc_pkcs15_cert_info_t * attcert_info = NULL;
+				sc_pkcs15_cert_t *attcert_out = NULL;
+				u8  *pp_val = NULL;
+				size_t pp_len = 0;
+				int pp_critical = 0;
+
+				attcert_info = calloc(1, sizeof(sc_pkcs15_cert_info_t));
+				if (attcert_info == NULL)
+					LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+
+				attcert_info->value.len = attcertctl.der.len;
+				attcert_info->value.value  = attcertctl.der.value;
+				attcertctl.der.value = NULL;
+
+				r = sc_pkcs15_read_certificate(p15card, attcert_info, &attcert_out);
+				if (r >= 0 && attcert_out && attcert_out->extensions) {
+					r = sc_pkcs15_get_extension(card->ctx, attcert_out,
+							&pp_oid, &pp_val, &pp_len, &pp_critical);
+					/* pp_val[0] pin policy default = 0, none = 1, session = 2, always = 3 */
+					/* pp_val[1] touch policy. TODO can we support this in some way */
+					if (r >= 0 && pp_len == 2 && pp_val) {
+						if (*pp_val != 0 && *pp_val != 3) {
+							prkey_obj.user_consent = 0; /* turn off */
+							sc_log(card->ctx, "user_consent turned off for 0x%2.2x", prkeys[i].ref);
```

Comment:
Would it be possible to evaluate the touch policy in the card driver? Everytime that key is used (and the user needs to touch), the card driver could send a notification so that the user is aware that he needs to do something. Similarily, [I've used `sc_notify()` in OpenSCTokenApp to tell the user to look at the PIN-pad of his reader](https://github.com/frankmorgner/OpenSCToken/blob/5359efaad81f2593526b7b30c3d423b840b21aa4/OpenSCToken/TokenSession.m#L128-L139). You could use the Yubikey's ATR to customize `MD_PINPAD_DLG_CONTENT_USER` and `MD_PINPAD_DLG_MAIN` with text for the touch button.

Reply:
Yes it could be done
My goal with this PR was to address the non standard PIV card's ability to set and enforce policies on the card for the 9c signing key.  I was not trying to address the touch policy or other keys. 

It could also be expanded to look at all keys and the "none" key policy. (PIV standard has this for the 9e key). 

(At my request, they sent me a YubiYey 5 NFC device with version 5.2.6 to use in testing of the Attestation certificate.)
Yubico is adding in 5.3 a simpler way to get the pin and touch policy without using the Attestation certificate. 

With Yubico devices (identified from their ATR) card-piv.c reads the Yubico version number to further distinguish differences in their devices. 

You could add this on the wish list.

Reply:
I'd like to avoid moving this from "not supported"" to "somewhat supported"; this only confuses users. For example, applications like MS Outlook are hanging during PIN verification. Adding a notification is the least we can do to get the user's attention so he has some clue of what he's supposed to do. I'd even go for a popup Window, but that would require even more work.